### PR TITLE
Feat/honeycomb scaler

### PR DESCRIPTION
**Pull request by kmoonwright**
## Honeycomb Scaler: Add New External Metrics Scaler

This PR introduces a new scaler for [Honeycomb.io](https://www.honeycomb.io), which enables KEDA to scale workloads based on observability data and queries executed against Honeycomb datasets.

### What Has Changed

- **New scaler implementation:**  
  Adds `scalers/honeycomb_scaler.go` (and associated test files) that allows users to scale workloads based on Honeycomb queries.
- **Metadata and config parsing:**  
  Supports both legacy config fields and flexible `queryRaw` JSON for arbitrary query construction.
- **API integration:**  
  Interacts with the Honeycomb API for query creation, execution, and polling.
- **Edge case handling:**  
  Handles response parsing, missing fields, non-numeric results, and timeouts robustly.
- **Unit tests:**  
  Comprehensive tests for metadata parsing and result field extraction.
- **CRD and deployment instructions:**  
  Updated CRD and deployment YAMLs to ensure compatibility with the new scaler.

---

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [ ] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [ ] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [ ] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [ ] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

---

**Fixes #<link main issue if exists>**  
Relates to #<link docs/helm PRs when opened>

---

**Notes:**  
- Please let me know if you’d like any adjustments to config or interface for the scaler before merging.
- I will open documentation and Helm chart PRs after initial review.

### CODE REVIEW

File: pkg/scalers/honeycomb_scaler.go [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+	}
+
+	// 2. Run Query
+	runURL := fmt.Sprintf("%s/query_results/%s", honeycombBaseURL, s.metadata.Dataset)
+	runBody, _ := json.Marshal(map[string]interface{}{
+		"query_id":                   createRes.ID,
+		"disable_series":             false,
+		"disable_total_by_aggregate": true,
+		"disable_other_by_aggregate": true,
+		"limit":                      10000,
```
Comment (dttung2905) [inline pkg/scalers/honeycomb_scaler.go]: Should we make this a constant ?

File: pkg/scalers/honeycomb_scaler.go [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+	})
+	runReq, _ := http.NewRequestWithContext(ctx, "POST", runURL, bytes.NewBuffer(runBody))
```
Comment (dttung2905) [inline pkg/scalers/honeycomb_scaler.go]: nit: can we handle the error returned from this ?

File: pkg/scalers/honeycomb_scaler.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+	for attempt := 0; attempt < maxPollAttempts; attempt++ {
+		time.Sleep(pollDelay)
+		pollDelay *= 2
+		statusReq, _ := http.NewRequestWithContext(ctx, "GET", pollURL, nil)
+		statusReq.Header.Set("X-Honeycomb-Team", s.metadata.APIKey)
+		statusResp, err := s.httpClient.Do(statusReq)
+		if err != nil {
+			return 0, fmt.Errorf("honeycomb poll query error: %w", err)
+		}
+		defer statusResp.Body.Close()
```
Comment (dttung2905) [inline pkg/scalers/honeycomb_scaler.go]: I dont thin this is a good idea as it might lead to resource leaks as these defer called will be accumulated inside the polling loop. 
Do you think we should `statusResp.Body.Close()` at the end of the loop ?

**Review by dttung2905 — COMMENTED** [review]
Hey @kmoonwright , thanks so much for working on this. I strongly believe this is a useful addition to the list of scaler we support. I know it is still a draft but i want to give some feedbacks/ nits as early as possible. Hope you don't mind :smile:

**Comment by kmoonwright** [comment]
Great feedback, thank you! Much appreciated even in the draft

**Review by zroubalik — COMMENTED** [review]
Hi @kmoonwright, thanks for the contribution.

Could you please also add e2e tests for this scaler? This is a requirement for us in order to proceed forward towards merging.

Thanks!

File: pkg/scalers/honeycomb_scaler.go [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+	}
+
+	// 2. Run Query
+	runURL := fmt.Sprintf("%s/query_results/%s", honeycombBaseURL, s.metadata.Dataset)
+	runBody, err := json.Marshal(map[string]interface{}{
+		"query_id":                   createRes.ID,
+		"disable_series":             false,
+		"disable_total_by_aggregate": true,
+		"disable_other_by_aggregate": true,
+		"limit":                      honeycombQueryResultsLimit,
```
[the remark this answers is not in the document]
Comment (kmoonwright) [reply pkg/scalers/honeycomb_scaler.go]: API query results are limited to 10000, comment added

File: pkg/scalers/honeycomb_scaler.go [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("error parsing honeycomb metadata: %w", err)
+	}
+
+	logger.Info("Initializing Honeycomb Scaler", "dataset", meta.Dataset)
+
+	return &honeycombScaler{
+		metricType: metricType,
+		metadata:   meta,
+		httpClient: &http.Client{Timeout: 15 * time.Second},
```
[the remark this answers is not in the document]
Comment (kmoonwright) [reply pkg/scalers/honeycomb_scaler.go:70]: Query Results cannot take longer than 10 seconds to run. Comment added.

File: pkg/scalers/honeycomb_scaler.go [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+		return 0, errors.New("runQuery: missing queryResult id")
+	}
+	if runRes.Complete && len(runRes.Data.Results) > 0 {
+		return extractResultField(runRes.Data.Results, s.metadata.ResultField)
+	}
+
+	// 3. Poll for completion (exponential backoff)
+	pollURL := fmt.Sprintf("%s/query_results/%s/%s", honeycombBaseURL, s.metadata.Dataset, runRes.ID)
+	pollDelay := initialPollDelay
+	for attempt := 0; attempt < maxPollAttempts; attempt++ {
```
[the remark this answers is not in the document]
Comment (kmoonwright) [reply pkg/scalers/honeycomb_scaler.go:204]: Polling is restricted Query Results API to 10 requests per minute, this configuration is to reasonably work within rate limits. Comment added.

**Comment by rickbrouwer** [comment]
/run-e2e honeycomb
**Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/16776705449)

File: CHANGELOG.md [hunk 7]
```diff
@@ -60,6 +60,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 
 ## Unreleased
 
+- **General**: Introduce new Honeycomb Scaler ([#6896](https://github.com/kedacore/keda/pull/6896/))
```
Comment (JorTurFer) [inline CHANGELOG.md:63]: Move this to `New` section

File: go.mod [hunk 8]
```diff
@@ -1,6 +1,6 @@
 module github.com/kedacore/keda/v2
 
-go 1.23.8
+go 1.23
```
Comment (JorTurFer) [inline go.mod:3]: Why is this required?

File: pkg/scalers/honeycomb_scaler.go [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+
+type honeycombScaler struct {
+	metricType v2.MetricTargetType
+	metadata   honeycombMetadata
+	httpClient *http.Client
+	logger     logr.Logger
+}
+
+type honeycombMetadata struct {
+	APIKey              string                 `keda:"name=apiKey, order=authParams;triggerMetadata"`
```
Comment (JorTurFer) [inline pkg/scalers/honeycomb_scaler.go:37]: We don't support reading secrets from metadata
```suggestion
	APIKey              string                 `keda:"name=apiKey, order=authParams"`
```

File: pkg/scalers/honeycomb_scaler.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (JorTurFer) [reply pkg/scalers/honeycomb_scaler.go:70]: Use the helper `CreateHTTPClient` to create the client. This adds the global configurations for HTTP Clients 
For example: https://github.com/kedacore/keda/blob/main/pkg/scalers/metrics_api_scaler.go#L105

File: pkg/scalers/honeycomb_scaler.go [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+func parseHoneycombMetadata(config *scalersconfig.ScalerConfig) (honeycombMetadata, error) {
+	meta := honeycombMetadata{}
+	err := config.TypedConfig(&meta)
+	if err != nil {
+		return meta, fmt.Errorf("error parsing honeycomb metadata: %w", err)
+	}
+	meta.TriggerIndex = config.TriggerIndex
+
+	// Use queryRaw if provided, else build query from legacy fields
+	if raw, ok := config.TriggerMetadata["queryRaw"]; ok && raw != "" {
```
Comment (JorTurFer) [inline pkg/scalers/honeycomb_scaler.go:84]: Can't we just use the parsed value?
```suggestion
	if meta.QueryRaw != "" {
```

File: pkg/scalers/honeycomb_scaler.go [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+
+func parseHoneycombMetadata(config *scalersconfig.ScalerConfig) (honeycombMetadata, error) {
+	meta := honeycombMetadata{}
+	err := config.TypedConfig(&meta)
+	if err != nil {
+		return meta, fmt.Errorf("error parsing honeycomb metadata: %w", err)
+	}
+	meta.TriggerIndex = config.TriggerIndex
+
+	// Use queryRaw if provided, else build query from legacy fields
```
Comment (JorTurFer) [inline pkg/scalers/honeycomb_scaler.go:83]: How many are options to set the config? Checking the code, user can set the config:
- using `queryRaw` to provide the values as json
- using `query` to provide values as map[string]interface (I'm not fully sure if this will work tbh)
- using `breakdowns`, `calculations`, etc
Am I right? If yes, I think that they are too many option and we should just choose one to avoid confusion to users. Are other softwares supporting those ways (maybe it's a common pattern for honeycomb) or it's just to bring all the options to users?

File: pkg/scalers/honeycomb_scaler.go [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+		}
+		meta.Query = q
+	}
+	if meta.Query == nil {
+		return meta, errors.New("no valid query provided in 'queryRaw', 'query', or legacy fields")
+	}
+	return meta, nil
+}
+
+func (s *honeycombScaler) Close(context.Context) error { return nil }
```
Comment (JorTurFer) [inline pkg/scalers/honeycomb_scaler.go:112]: We should close idle connections on closing
```suggestion
func (s *honeycombScaler) Close(context.Context) error {
	if s.httpClient != nil {
		s.httpClient.CloseIdleConnections()
	}
	return nil
}
```

File: pkg/scalers/honeycomb_scaler.go [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+		return 0, fmt.Errorf("error creating Honeycomb create query request: %w", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("X-Honeycomb-Team", s.metadata.APIKey)
+	resp, err := s.httpClient.Do(req)
+	if err != nil {
+		return 0, fmt.Errorf("honeycomb create query error: %w", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 && resp.StatusCode != 201 {
```
Comment (JorTurFer) [inline pkg/scalers/honeycomb_scaler.go:134]: Does this API returns 200 and 201? I'd expect just only one of them

File: pkg/scalers/honeycomb_scaler.go [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+	}
+	return meta, nil
+}
+
+func (s *honeycombScaler) Close(context.Context) error { return nil }
+
+// ----- Core Query Logic -----
+
+func (s *honeycombScaler) executeHoneycombQuery(ctx context.Context) (float64, error) {
+	// 1. Create Query
```
Comment (JorTurFer) [inline pkg/scalers/honeycomb_scaler.go:117]: Can we create a query on each request? I mean, we will create multiple queries per minute, is this correct from honeycomb pov or users should create the query by their own somewhere and provide the query ID as parameter?

File: pkg/scalers/honeycomb_scaler.go [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+
+	// 2. Run Query
+	runURL := fmt.Sprintf("%s/query_results/%s", honeycombBaseURL, s.metadata.Dataset)
+	runBody, err := json.Marshal(map[string]interface{}{
+		"query_id":                   createRes.ID,
+		"disable_series":             false,
+		"disable_total_by_aggregate": true,
+		"disable_other_by_aggregate": true,
+		// Query results limit is 10000, see https://api-docs.honeycomb.io/api for details
+		"limit": honeycombQueryResultsLimit,
```
Comment (JorTurFer) [inline pkg/scalers/honeycomb_scaler.go:156]: Does it make sense to expose this to users? I mean, the limit must be enforced, but maybe a user prefers to query X items instead of the limit. Maybe we can expose this to users and check that value doesn't pass the limit.
Or maybe this is not a valid use case, saying the truth, I don't have any experience with honeycomb

File: pkg/scalers/honeycomb_scaler.go [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+	}
+	defer runResp.Body.Close()
+	if runResp.StatusCode == 429 {
+		return 0, errors.New("honeycomb: rate limited (429), back off and try again later")
+	}
+	if runResp.StatusCode == 401 {
+		body, _ := io.ReadAll(runResp.Body)
+		return 0, fmt.Errorf("honeycomb: unauthorized (401) - an Enterprise API key is required, check your API key permissions. See: https://api-docs.honeycomb.io/api/query-data for details. Response: %s", string(body))
+	}
+	if runResp.StatusCode != 200 && runResp.StatusCode != 201 {
```
Comment (JorTurFer) [inline pkg/scalers/honeycomb_scaler.go:179]: Same as above

File: pkg/scalers/honeycomb_scaler.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (JorTurFer) [reply pkg/scalers/honeycomb_scaler.go:204]: as default, a single metric requires 6 queries per minute if everything works at first try (every 15 seconds the HPA controller checks the value and every 30 seconds the operator). If it requires 2 checks per metric, the rate limit will be broken with a single metric. Is the rate limit applied by API or by queryID?

File: pkg/scalers/honeycomb_scaler.go [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+		statusReq.Header.Set("X-Honeycomb-Team", s.metadata.APIKey)
+		statusResp, err := s.httpClient.Do(statusReq)
+		if err != nil {
+			return 0, fmt.Errorf("honeycomb poll query error: %w", err)
+		}
+		if statusResp.StatusCode == 429 {
+			statusResp.Body.Close()
+			return 0, errors.New("honeycomb: rate limited (429) on poll, back off and try again later")
+		}
+		if statusResp.StatusCode != 200 && statusResp.StatusCode != 201 {
```
Comment (JorTurFer) [inline pkg/scalers/honeycomb_scaler.go:220]: same as above

File: tests/scalers/honeycomb/honeycomb_test.go [hunk 18]
```diff
@@ -0,0 +1,308 @@
+//go:build e2e
```
Comment (JorTurFer) [inline tests/scalers/honeycomb/honeycomb_test.go:1]: How are the metrics sent to honeycomb? I don't see any agent or so that collects them

**Review by zroubalik — COMMENTED** [review]
@kmoonwright any update here please?
